### PR TITLE
Add lockup duration edit method

### PR DESCRIPTION
**Pull request by mconcat**
## Description

Closes: #164

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by ValarDragon** [comment]
Ah sweet! Lets add a hook for this?

Also thoughts on renaming `EditLockup` to `ExtendLockup`?

**Comment by mconcat** [comment]
> Ah sweet! Lets add a hook for this?
> 
> Also thoughts on renaming `EditLockup` to `ExtendLockup`?

I was thinking about using this method for editing other part of the lockups in the future(e.g. cancelling unlock) so we can reduce the total number of messages. If you think they should be in different message, I agree ExtendLockup is a better name!

**Comment by ValarDragon** [comment]
I think they should be different messages!

(It is an interesting API question if CancelUnlock should be expected behavior of extend lockup on an unbonding lockup, but I think thats a future question)

File: proto/osmosis/lockup/tx.proto
Comment (alexanderbez) [inline proto/osmosis/lockup/tx.proto:55]: nit: protodoc

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:542]:
```suggestion
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:544]:
```suggestion
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:555]:
```suggestion
```

File: x/lockup/keeper/msg_server.go
Comment (alexanderbez) [inline x/lockup/keeper/msg_server.go:168]: Nit: Use `Wrapf` instead of `Sprintf`

File: x/lockup/types/msgs.go
Comment (alexanderbez) [inline x/lockup/types/msgs.go:109]: There's zero validation we want to do here? What about checking the fields for basic bounds?

File: x/lockup/keeper/lock.go
Comment (mattverse) [inline x/lockup/keeper/lock.go:570]: Do we not need to run `deleteLockRefs` to delete old lock refs? I'm not quite sure if `setLockAndResetLockRefs` is actually deleting the old refs

File: x/lockup/types/msgs.go
Comment (mconcat) [reply x/lockup/types/msgs.go:109]: Yup seems like adding nil/empty checks will be needed!

File: x/lockup/keeper/lock.go
Comment (mconcat) [reply x/lockup/keeper/lock.go:570]: Thanks for catching it! Will fix this

Comment (mconcat) [reply x/lockup/keeper/lock.go:570]: Matt pointed out that `setLockAndResetLockRefs` is bit misleading, because the name and the comment
```
// setLockAndResetLockRefs sets the lock, and resets all of its lock references
// This puts the lock into a 'clean' state, aside from the AccumulationStore.
```
implies that it removes the existing lockrefs, but actually it doesn't. Should we fix this behaviour, or change the name of the method?

Comment (mattverse) [reply x/lockup/keeper/lock.go]: I think we wouldn't need to fix this within this PR, but instead create an issue to investigate on this.

That being said, I think it would suffice for this PR to ensure that old lock refs are getting deleted using `deleteLockRefs`

Comment (mconcat) [reply x/lockup/keeper/lock.go]: I have replaced setLockAnd.. to a pair of delete and add lockrefs. This works!

Comment (ValarDragon) [reply x/lockup/keeper/lock.go]: Oh interesting, lets make an issue and investigate this

Comment (mattverse) [reply x/lockup/keeper/lock.go]: https://github.com/osmosis-labs/osmosis/issues/1275

File: x/lockup/types/msgs.go
Comment (alexanderbez) [reply x/lockup/types/msgs.go:112]: What about making sure the owner is empty?

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:543]:
```suggestion
		return fmt.Errorf("cannot edit unlocking lockup for lock %d", lock.ID)
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:548]:
```suggestion
		return fmt.Errorf("cannot edit lockup with synthetic lock %d", lock.ID)
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:554]:
```suggestion
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:555]:
```suggestion
		if newDuration < lock.Duration {
```

File: x/lockup/keeper/lock.go
Comment (alexanderbez) [inline x/lockup/keeper/lock.go:577]:
```suggestion
	// update lockup
	err := k.deleteLockRefs(ctx, unlockingPrefix(oldLock.IsUnlocking()), oldLock)
	if err != nil {
		return err
	}

	err = k.addLockRefs(ctx, lock)
	if err != nil {
		return err
	}

	return k.setLock(ctx, lock)
```

**Comment by stackman27** [comment]
Closes [#164](https://github.com/osmosis-labs/osmosis/issues/164) ?

**Review by alexanderbez — APPROVED** [review]

**Review by alexanderbez — APPROVED** [review]

**Comment by alexanderbez** [comment]
@mattverse wanna give this another look?

File: x/lockup/keeper/msg_server.go
Comment (mattverse) [inline x/lockup/keeper/msg_server.go:185]: We should either fill in the response field or delete it in the proto file if it's unused

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [inline x/lockup/keeper/lock_test.go:679]: Can we add basic tests to ensure other fields of lock itself is preserved after `ExtendLockup`?
```suggestion
	suite.Require().Equal(1, lock.ID)
	suite.Require().Equal(coins, lock.Coins)
	
```

File: x/lockup/keeper/lock_test.go
Comment (mattverse) [inline x/lockup/keeper/lock_test.go:671]: Could we also create test cases to err if the lock has same duration as the original lock?
```suggestion
	// extending lock with same duration should fail
	err = suite.App.LockupKeeper.ExtendLockup(suite.Ctx, *lock, time.Second)
	suite.Require().Error(err)
	
```

**Review by mattverse — COMMENTED** [review]
Don't want to block this PR for so long being picky 😅 

Went through final review, should be good to merge after fix!

**Review by mattverse — APPROVED** [review]
🌮

File: proto/osmosis/lockup/tx.proto
Comment (ValarDragon) [inline proto/osmosis/lockup/tx.proto:72]: Should we remove this response field? (Can be done in a second PR)

Not sure what this would communicate relative to the general txs success / failure

File: x/lockup/keeper/lock.go
Comment (ValarDragon) [inline x/lockup/keeper/lock.go:584]: Lets re-assess these hook calls in a follow-up issue

File: x/lockup/keeper/msg_server_test.go
Comment (ValarDragon) [inline x/lockup/keeper/msg_server_test.go:266]: Great job on this test!

**Review by ValarDragon — APPROVED** [review]
Nice work on this!

**Pull request merged**
